File: src/test/java/pcoet/patterns/data/structures/BagTest.java
package pcoet.patterns.data.structures;

import static org.hamcrest.CoreMatchers.equalTo;
import static org.hamcrest.CoreMatchers.not;
import static org.junit.Assert.assertThat;

import org.junit.Test;

import java.util.Iterator;

public class BagTest {
  Bag<String> bag;

  private void setUpStack() {
    bag = new Bag<>();
    bag.add("lions");
    bag.add("tigers");
    bag.add("bears");
  }

  public BagTest() {
    this.setUpStack();
  }

  @Test
  public void bagWithItemsIsNotEmpty() {
    assertThat(bag.isEmpty(), not(true));
  }

  @Test
  public void bagHasExpectedSize() {
    assertThat(bag.size(), equalTo(3));
  }

  @Test
  public void canIterateOverItems() {
    Iterator<String> iterator = bag.iterator();
    assertThat(iterator.hasNext(), equalTo(true));
    assertThat(iterator.next(), equalTo("bears"));
    assertThat(iterator.next(), equalTo("tigers"));
    assertThat(iterator.next(), equalTo("lions"));
  }
}

File: app/src/test/kotlin/com/pcoet/patterns/data/structures/BagTest.kt
package com.pcoet.patterns.data.structures

import org.junit.Test

import org.junit.Assert.*

class BagTest {

  @Test
  fun isEmpty() {
    val bag = Bag<Int>()
    assertEquals(true, bag.isEmpty())
    bag.add(1)
    assertEquals(false, bag.isEmpty())
  }

  @Test
  fun size() {
    val bag = Bag<String>()
    assertEquals(0, bag.size())
    bag.add("abc")
    assertEquals(1, bag.size())
  }

  @Test
  fun add() {
    val bag = Bag<Int>()
    bag.add(1).add(1).add(2).add(3).add(5).add(8).add(13)
    assertEquals(7, bag.size())
  }

  /**
   * For simplicity, this test assumes an order to the iteration, but that is an implementation detail.
   */
  @Test
  operator fun iterator() {
    val fibs = listOf(1, 1, 2, 3, 5, 8, 13)
    val bag = Bag<Int>().add(13).add(8).add(5).add(3).add(2).add(1).add(1)
    var i = 0
    bag.forEach {
      assertEquals(fibs[i], it)
      i++
    }
  }
}